src/y0/algorithm/transport.py

```diff
@@ -0,0 +1,464 @@
+"""Implement of surrogate outcomes and transportability from https://arxiv.org/abs/1806.07172."""
+
+import logging
+from copy import deepcopy
+from dataclasses import dataclass
+from operator import attrgetter
+from typing import Dict, FrozenSet, Iterable, List, Mapping, Optional, Set, Union, cast
+
+from y0.algorithm.conditional_independencies import are_d_separated
+from y0.dsl import (
+    CounterfactualVariable,
+    Expression,
+    Intervention,
+    One,
+    Population,
+    Product,
+    Sum,
+    Variable,
+)
+from y0.graph import NxMixedGraph
+
+__all__ = [
+    "transport",
+    "TransportQuery",
+]
+
+logger = logging.getLogger(__name__)
+
+TARGET_DOMAIN = Population("pi*")
+
+
+def get_nodes_to_transport(
+    *,
+    surrogate_interventions: Union[Set[Variable], Variable],
+    surrogate_outcomes: Union[Set[Variable], Variable],
+    graph: NxMixedGraph,
+) -> Set[Variable]:
+    """Identify which nodes the transport nodes should point to.
+
+    :param surrogate_interventions: The interventions performed in an experiment.
+    :param surrogate_outcomes: The outcomes observed in an experiment.
+    :param graph: The graph of the target domain.
+    :returns: A set of variables representing target domain nodes where transportability nodes should be added.
+    """
+    if isinstance(surrogate_interventions, Variable):
+        surrogate_interventions = {surrogate_interventions}
+    if isinstance(surrogate_outcomes, Variable):
+        surrogate_outcomes = {surrogate_outcomes}
+
+    # Find the c_component with surrogate_outcomes
```

Comment:
is there a case for this being a more generic function for finding the "c-component" for multiple nodes?

We can also implement this upside down:

```python
    c_component_surrogate_outcomes: Set[Variable] = set()
    for o in surrogate_outcomes:
        c_component_surrogate_outcomes.update(graph.get_district(outcome))
```